src/validators/string.rs

```diff
@@ -230,18 +230,48 @@ impl RegexEngine {
 }
 
 impl Pattern {
-    fn compile(py: Python<'_>, pattern: String, engine: &str) -> PyResult<Self> {
+    fn extract_pattern_str(pattern: Bound<'_, PyAny>) -> PyResult<String> {
+        if pattern.is_instance_of::<PyString>() {
+            Ok(pattern.to_string())
+        } else {
+            match pattern.getattr("pattern") {
+                Ok(attr) => attr.extract::<String>(),
+                Err(_) => Err(py_schema_error_type!(
+                    "Invalid pattern, must be str or re.Pattern: {}",
+                    pattern
+                )),
+            }
+        }
+    }
+
+    fn compile_pattern(pattern: Bound<'_, PyAny>) -> PyResult<PyObject> {
+        let py = pattern.py();
+
+        let re_module = py.import_bound(intern!(py, "re"))?;
+        let re_compile = re_module.getattr(intern!(py, "compile"))?;
+        let re_pattern = re_module.getattr(intern!(py, "Pattern"))?;
+
+        if pattern.is_instance(&re_pattern)? {
+            Ok(pattern.to_object(py))
+        } else {
+            Ok(re_compile.call1((pattern,))?.into())
+        }
+    }
+
+    fn compile(pattern: Bound<'_, PyAny>, engine: &str) -> PyResult<Self> {
+        let pattern_str = Self::extract_pattern_str(pattern.clone())?;
+
         let engine = match engine {
             RegexEngine::RUST_REGEX => {
-                RegexEngine::RustRegex(Regex::new(&pattern).map_err(|e| py_schema_error_type!("{}", e))?)
-            }
-            RegexEngine::PYTHON_RE => {
-                let re_compile = py.import_bound(intern!(py, "re"))?.getattr(intern!(py, "compile"))?;
-                RegexEngine::PythonRe(re_compile.call1((&pattern,))?.into())
+                RegexEngine::RustRegex(Regex::new(&pattern_str).map_err(|e| py_schema_error_type!("{}", e))?)
```

Comment:
I think we should go the other way around here - if it's a `re.Pattern` object then we should ignore whether the user has requested Python or Rust engine and instead just make a `PythonRe` engine which wraps the pattern. I think it's easy enough to document this; then the regex engine field is only used if basic strings are passed.

Otherwise we're just moving the original bug around and dropping the flags when we build the Rust engine 🙈